src/EditorFeatures/Core/Tagging/TaggerMainThreadManager.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+internal sealed class TaggerMainThreadManager
+{
+    private static readonly ConditionalWeakTable<IThreadingContext, TaggerMainThreadManager> s_table = new();
+
+    public static TaggerMainThreadManager GetManager(IThreadingContext threadingContext, IAsynchronousOperationListenerProvider listenerProvider)
+        => s_table.GetValue(threadingContext, _ => new TaggerMainThreadManager(threadingContext, listenerProvider));
```

Comment:
>s_table [](http://example.com/codeflow?start=11&length=7)

can you comment why we need a CWT? In my mind, there is a single IThreadingContext, but there must be cases where there are multiple?

Reply:
There is one per composition. But we have tons of compositions.  For example, in tests we make a lot of them. 

When the composition is torn down (which tests do), the disposal token is triggered.  Which would kill this type. 

That said, we can make the TaggerMainThreadManager. A mef component.  Then it can import its own IThreadingContext and async listener provider.  

One benefit is that this works with all taggers (including TS ones).